torch/csrc/deploy/deploy.h

```diff
@@ -0,0 +1,235 @@
+#pragma once
+#include <assert.h>
+#include <torch/csrc/deploy/interpreter/interpreter_impl.h>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace torch {
+
+struct MovableObject;
+struct InterpreterManager;
+
+struct TORCH_API InterpreterSession {
+  InterpreterSession(InterpreterSessionImpl* impl, InterpreterManager* manager)
+      : impl_(impl), manager_(manager) {}
+
+  PythonObject self; // when retreived from a PythonMovable this will be set.
+  InterpreterSession(InterpreterSession&&) = default;
+  ~InterpreterSession();
+  PythonObject global(const char* module, const char* name) {
+    return impl_->global(module, name);
+  }
+  PythonObject from_ivalue(at::IValue ivalue) {
+    return impl_->from_ivalue(std::move(ivalue));
+  }
+
+  MovableObject create_movable(PythonObject obj);
+
+ private:
+  friend struct MovableObject;
+  friend struct Package;
+  friend struct InterpreterManager;
+  friend struct MovableObjectImpl;
+  std::unique_ptr<InterpreterSessionImpl> impl_;
+  InterpreterManager* manager_; // if created from one
+  int64_t notify_idx_ = -1;
+};
+
+class TORCH_API Interpreter {
+ private:
+  std::string library_name_;
+  void* handle_;
+  std::unique_ptr<InterpreterImpl> pImpl_;
+
+  InterpreterManager* manager_; // optional if managed by one
+
+ public:
+  Interpreter(InterpreterManager* manager);
+  InterpreterSession acquire_session() const {
+    return InterpreterSession(pImpl_->acquire_session(), manager_);
+  }
+  ~Interpreter();
+  Interpreter(Interpreter&& rhs)
+      : library_name_(std::move(rhs.library_name_)),
+        handle_(rhs.handle_),
+        pImpl_(std::move(rhs.pImpl_)),
+        manager_(rhs.manager_) {
+    rhs.handle_ = nullptr;
+  }
+
+  Interpreter(const Interpreter&) = delete;
+  Interpreter& operator=(const Interpreter&) = delete;
+  Interpreter& operator=(Interpreter&&) = delete;
+  friend struct InterpreterManager;
+};
+
+struct Package;
+
+struct TORCH_API LoadBalancer {
```

Comment:
This is what I ended up with after timing scaling of an effectively no-op model. With 64 threads, anything that tries to take locks doesn't scale. Overkill for not very small models, but this logic is pretty simply anyway.

Reply:
I wonder whether in real setup we'd just have one thread per interpreter so then implementation would be to have an efficient queue of requests and each thread trying to pull from it (with something like folly::MPMCQueue making it very efficient). Or do you think overheads per-interpreter are high enough (mostly memory?) to justify sharing interpreters across threads? Even in that case, we can still have just fixed # of threads per interpreter.

Reply:
The intent is for this to be very fast when n_hardware_threads == n_interpreters, but for it to still work correctly in other conditions. This design means that clients of the library don't have to worry about pinning interpreters to threads, or even really realizing that multiple interpreters are involved in the first place. 

We need to avoid having n_interpreters == n_software_threads. Each one of these interpreters takes ~1--2 seconds to startup because its dlopning our python bindings and `import torch`ing. At startup we want to just create enough interpreters to cover our cpu cores and then never have to create another interpreter.